src/libproc_macro/lib.rs

```diff
@@ -69,6 +76,443 @@ pub struct LexError {
+    }
+}
+
+#[unstable(feature = "proc_macro", issue = "38356")]
+impl ops::Deref for Symbol {
+    type Target = str;
+
+    fn deref(&self) -> &str {
+        unsafe { &*(self.0.as_str().deref() as *const str) }
+    }
+}
+
+/// Whether an `Op` is either followed immediately by another `Op` or followed by whitespace.
+#[derive(Copy, Clone)]
+#[unstable(feature = "proc_macro", issue = "38356")]
+pub enum OpKind {
+    /// e.g. `+` is `Alone` in `+ =`.
+    Alone,
+    /// e.g. `+` is `Joint` in `+=`.
+    Joint,
+}
+
+/// A literal character (`'a'`), string (`"hello"`), or number (`2.3`).
+#[derive(Clone)]
+#[unstable(feature = "proc_macro", issue = "38356")]
+pub struct Literal(token::Token);
+
+#[unstable(feature = "proc_macro", issue = "38356")]
+impl fmt::Display for Literal {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        TokenTree { kind: TokenKind::Literal(self.clone()), span: Span(DUMMY_SP) }.fmt(f)
+    }
+}
+
+macro_rules! int_literals {
+    ($($int_kind:ident),*) => {$(
+        /// Integer literal.
+        #[unstable(feature = "proc_macro", issue = "38356")]
+        pub fn $int_kind(n: $int_kind) -> Literal {
+            Literal::integer(n as i128, stringify!($int_kind))
+        }
+    )*}
+}
+
+impl Literal {
+    int_literals!(u8, i8, u16, i16, u32, i32, u64, i64);
+    fn integer(n: i128, kind: &'static str) -> Literal {
+        Literal(token::Literal(token::Lit::Integer(symbol::Symbol::intern(&n.to_string())),
+                               Some(symbol::Symbol::intern(kind))))
+    }
+
+    /// Floating point literal.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn f32(n: f32) -> Literal {
+        Literal(token::Literal(token::Lit::Float(symbol::Symbol::intern(&n.to_string())),
+                               Some(symbol::Symbol::intern("f32"))))
+    }
+
+    /// Floating point literal.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn f64(n: f32) -> Literal {
+        Literal(token::Literal(token::Lit::Float(symbol::Symbol::intern(&n.to_string())),
+                               Some(symbol::Symbol::intern("f64"))))
+    }
+
+    /// String literal.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn string(string: &str) -> Literal {
+        let mut escaped = String::new();
+        for ch in string.chars() {
+            escaped.extend(ch.escape_unicode());
+        }
+        Literal(token::Literal(token::Lit::Str_(symbol::Symbol::intern(&escaped)), None))
+    }
+
+    /// Character literal.
+    #[unstable(feature = "proc_macro", issue = "38356")]
+    pub fn character(ch: char) -> Literal {
+        let mut escaped = String::new();
+        escaped.extend(ch.escape_unicode());
+        Literal(token::Literal(token::Lit::Char(symbol::Symbol::intern(&escaped)), None))
+    }
+}
+
+/// An iterator over `TokenTree`s.
+#[unstable(feature = "proc_macro", issue = "38356")]
+pub struct TokenIter {
```

Comment:
@nrc Yeah, `TokenTreeIter` makes more sense.

On a similar note, @eddyb proposed using `TokenNode` instead of `TokenKind` for better correspondence to `TokenTree` and to leave room for `Token`/`TokenKind` -- what do you think?